src/worker.cc

```diff
@@ -0,0 +1,694 @@
+                        worker_event_loop,
+                        argc,
+                        const_cast<const char**>(argv),
+                        exec_argc,
+                        const_cast<const char**>(exec_argv),
+                        keep_alive);
+
+  err = uv_thread_create(context->worker_thread(), RunWorkerInstance, context);
+
+  if (err != 0) {
+    context->worker_notifications()->Dispose();
+    context->Dispose();
+    CHECK_EQ(uv_loop_close(worker_event_loop), 0);
+    free(worker_event_loop);
+    return env->ThrowUVException(err, "uv_thread_create");
+  }
+}
+
+void WorkerContext::PostMessageToOwner(WorkerMessage* message) {
+  if (termination_kind() != NONE) {
+    delete message;
+    return;
+  }
+
+  if (!to_owner_messages_primary_.PushBack(message)) {
+    ScopedLock::Mutex lock(to_owner_messages_mutex());
+    to_owner_messages_backup_.PushBack(message);
+  }
+  owner_notifications()->Notify();
+}
+
+void WorkerContext::PostMessageToOwner(
+    const FunctionCallbackInfo<Value>& args) {
+  CHECK_EQ(args.Length(), 3);
+  WorkerContext* context = node::Unwrap<WorkerContext>(args.Holder());
+  if (context == nullptr)
+    return;
+  CHECK_CALLED_FROM_WORKER(context);
+  CHECK(args[0]->IsString());
+  // transferList ignored for now
+  CHECK(args[1]->IsArray());
+  CHECK(args[2]->IsInt32());
+  Local<String> json = args[0].As<String>();
+  WorkerMessageType message_type =
+      static_cast<WorkerMessageType>(args[2]->Int32Value());
+  WorkerMessage* message =
+      new WorkerMessage(ToUtf8Value(context->worker_env()->isolate(), json),
+                        message_type);
+  context->PostMessageToOwner(message);
+}
+
+void WorkerContext::PostMessageToWorker(WorkerMessage* message) {
+  if (termination_kind() != NONE) {
+    delete message;
+    return;
+  }
+
+  if (!to_worker_messages_primary_.PushBack(message)) {
+    ScopedLock::Mutex lock(to_worker_messages_mutex());
+    to_worker_messages_backup_.PushBack(message);
+  }
+
+  worker_notifications()->Notify();
+}
+
+void WorkerContext::PostMessageToWorker(
+    const FunctionCallbackInfo<Value>& args) {
+  WorkerContext* context = node::Unwrap<WorkerContext>(args.Holder());
+  if (context == nullptr)
+    return;
+  CHECK_CALLED_FROM_OWNER(context);
+  CHECK(args[0]->IsString());
+  // transferList ignored for now
+  CHECK(args[1]->IsArray());
+  CHECK(args[2]->IsInt32());
+  Local<String> json = args[0].As<String>();
+  WorkerMessageType message_type =
+      static_cast<WorkerMessageType>(args[2]->Int32Value());
+  Environment* env = Environment::GetCurrent(args);
+  WorkerMessage* message = new WorkerMessage(ToUtf8Value(env->isolate(),
+                                                         json),
+                                             message_type);
+  context->PostMessageToWorker(message);
+}
+
+void WorkerContext::WrapperNew(const FunctionCallbackInfo<Value>& args) {
+  HandleScope handle_scope(args.GetIsolate());
```

Comment:
If this is being called from JS then no need for the additional `HanldeScope`. There's an implicit one when calling from JS.